src/token/cose/key.rs

```diff
@@ -0,0 +1,444 @@
+    fn from(value: CoseEc2Key<'a, OE>) -> Self {
+        CoseParsedKey::Ec2(value)
+    }
+}
+
+impl<'a, OE: Display> From<CoseOkpKey<'a, OE>> for CoseParsedKey<'a, OE> {
+    fn from(value: CoseOkpKey<'a, OE>) -> Self {
+        CoseParsedKey::Okp(value)
+    }
+}
+
+impl<'a, OE: Display> From<CoseSymmetricKey<'a, OE>> for CoseParsedKey<'a, OE> {
+    fn from(value: CoseSymmetricKey<'a, OE>) -> Self {
+        CoseParsedKey::Symmetric(value)
+    }
+}
+
+pub type EllipticCurve = RegisteredLabelWithPrivate<iana::EllipticCurve>;
+
+pub struct CoseEc2Key<'a, OE: Display> {
+    generic: &'a CoseKey,
+    pub crv: EllipticCurve,
+    pub d: Option<&'a [u8]>,
+    pub x: Option<&'a [u8]>,
+    pub y: Option<&'a [u8]>,
+    pub sign: Option<bool>,
+    _backend_error_type: PhantomData<OE>,
+}
+
+impl<'a, OE: Display> TryFrom<&'a CoseKey> for CoseEc2Key<'a, OE> {
+    type Error = CoseCipherError<OE>;
+
+    fn try_from(key: &'a CoseKey) -> Result<Self, Self::Error> {
+        // Unless stated otherwise, these checks are according to RFC 9053, Section 7.1.1.
+
+        // Curve must be set
+        let crv = find_param_by_label(
+            &Label::Int(iana::Ec2KeyParameter::Crv.to_i64()),
+            &key.params,
+        )
+        .ok_or(CoseCipherError::MissingKeyParam(
+            iana::Ec2KeyParameter::Crv.into(),
+        ))?;
+        // Curve must be of correct type
+        let crv = EllipticCurve::from_cbor_value(crv.clone()).map_err(|e| {
+            CoseCipherError::InvalidKeyParam(iana::Ec2KeyParameter::Crv.into(), crv.clone())
+        })?;
+
+        // Check whether curve and key type are consistent (RFC 9053, Section 7.1)
+        match &crv {
+            EllipticCurve::Assigned(
+                iana::EllipticCurve::P_256
+                | iana::EllipticCurve::P_384
+                | iana::EllipticCurve::P_521
+                | iana::EllipticCurve::Secp256k1,
+                // TODO these are also part of the IANA registry, and should be added to coset
+                // | iana::EllipticCurve::BrainpoolP256r1
+                // | iana::EllipticCurve::BrainpoolP320r1
+                // | iana::EllipticCurve::BrainpoolP384r1
+                // | iana::EllipticCurve::BrainpoolP512r1
+            ) => {}
+            // Any other assigned values must not be encoded as EC2 values.
+            v @ EllipticCurve::Assigned(_) => {
+                return Err(CoseCipherError::KeyTypeCurveMismatch(
+                    KeyType::Assigned(iana::KeyType::EC2),
+                    v.clone(),
+                ))
+            }
+            // Anything else is private use/custom, and we assume that the caller knows what they
+            // are doing.
+            _v => {}
+        }
+
+        // Parse parameters d and x (private key and part of the public key), must be of type
+        // bstr.
+        let d = find_param_by_label(&Label::Int(iana::Ec2KeyParameter::D.to_i64()), &key.params)
```

Comment:
Haven't reviewed the new code yet, but if we can't always expect a sorted list here, maybe we can just do a linear search instead? Otherwise, the benefits of binary searching might be outweighed by the required sorting step (depending on how often we have to call `find_param_by_label`).

Reply:
The way it is now implemented the sorting function will only be called once per parsing method.

However, linear search might still be faster and less error-prone here (as the list of elements should be rather small), so I'll consider replacing it anyways.